// @vsintellicode/vscode-intellicode-python/build/vscode-intellicode-python.js
"use strict";
/*! Copyright (c) Microsoft Corporation. All rights reserved. */
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const vscode = __importStar(require("vscode"));
const path_1 = __importDefault(require("path"));
const fs_1 = __importDefault(require("fs"));
const genericErrorMessage = "Cannot start IntelliCode support for Python. See output window for more details.";
const minModelSchemaVersion = 2;
const maxModelSchemaVersion = 2;
class PythonSupport {
    getRequestedConfig() {
        return [{
                scopeName: "python",
                settingName: "jediEnabled",
                desiredValue: false,
                required: true,
                scopesToTry: [
                    vscode.ConfigurationTarget.Global,
                    vscode.ConfigurationTarget.Workspace,
                    vscode.ConfigurationTarget.WorkspaceFolder
                ],
                reloadWindowAfterApplying: true,
                notificationMessage: "IntelliCode Python support requires you to use the Microsoft Python Language Server (preview).",
                actionLabel: "Enable it and Reload Window"
            }];
    }
    async activate(api, logger) {
        let model = await api.ModelAcquisitionService
            .getModelProvider("PythonIntellicode", "Completions", minModelSchemaVersion, maxModelSchemaVersion, [])
            .getModelAsync(undefined);
        if (model === undefined) {
            logger("No model available for Python, cannot continue.");
            return Promise.resolve();
        }
        let modelJson = JSON.stringify(model);
        logger(`vs-intellicode-python was passed a model: ${modelJson}.`);
        let assemblyPath = path_1.default.join(__dirname, "IntellicodeForPython.dll");
        try {
            fs_1.default.accessSync(assemblyPath, fs_1.default.constants.F_OK);
        }
        catch (err) {
            logger(`Python Language Server extension assembly doesn't exist in ${assemblyPath}. Please reinstall IntelliCode.`);
            return Promise.reject(err);
        }
        const pythonExtension = vscode.extensions.getExtension("ms-python.python");
        if (!pythonExtension) {
            const err = "Microsoft Python extension is not installed.";
            logger(err);
            return Promise.reject(err);
        }
        if (!pythonExtension.isActive) {
            const api = await pythonExtension.activate();
            await api.ready;
        }
        let command = vscode.commands.executeCommand("python._loadLanguageServerExtension", {
            assembly: assemblyPath,
            typeName: "Microsoft.PythonTools.Analysis.Pythia.LanguageServerExtensionProvider",
            properties: {
                modelPath: model.modelPath
            }
        });
        if (command == null) {
            logger("Couldn't find language server extension command. Is the installed version of Python 2018.7.0 or later?");
            return Promise.reject(new Error(genericErrorMessage));
        }
        await command;
        logger("Loaded language server extension.");
        return Promise.resolve();
    }
}
exports.PythonSupport = PythonSupport;
