// tsdi/dist/tests/user.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
var _1 = require("../lib/");
var dependency_1 = require("./dependency");
var User = /** @class */ (function () {
    function User() {
    }
    User.prototype.getDep = function () {
        return this.dependency;
    };
    User.prototype.init = function () {
        this.initMessage = 'init';
    };
    User.prototype.initResult = function () {
        return this.initMessage;
    };
    User.prototype.method = function () {
        return this.dependency.echo('hello');
    };
    __decorate([
        _1.Inject(),
        __metadata("design:type", dependency_1.Dependency)
    ], User.prototype, "dependency", void 0);
    __decorate([
        _1.Initialize(),
        __metadata("design:type", Function),
        __metadata("design:paramtypes", []),
        __metadata("design:returntype", void 0)
    ], User.prototype, "init", null);
    User = __decorate([
        _1.Component()
    ], User);
    return User;
}());
exports.User = User;
