// tsdi/dist/lib/factory.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var global_state_1 = require("./global-state");
var debug = require("debug");
var log = debug('tsdi');
function Factory() {
    var args = [];
    for (var _i = 0; _i < arguments.length; _i++) {
        args[_i] = arguments[_i];
    }
    var decorate = function (target, propertyKey, options) {
        if (log.enabled) {
            log('@Factory %s#%s({name: "%s"})', target.constructor.name, propertyKey, target[propertyKey].name);
        }
        global_state_1.addKnownComponent({
            target: target,
            property: propertyKey.toString(),
            options: options,
            rtti: Reflect.getMetadata('design:returntype', target, propertyKey)
        });
    };
    if (args.length > 1) {
        return decorate(args[0], args[1], {});
    }
    var options = args[0] || {};
    return function (target, propertyKey) {
        decorate(target, propertyKey, options);
    };
}
exports.Factory = Factory;
exports.factory = Factory;
