// tsdi/dist/tests/mock-test.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
var chai_1 = require("chai");
require("source-map-support/register");
var tsdi_1 = require("../lib/tsdi");
describe('TSDI', function () {
    var tsdi;
    beforeEach(function () {
        tsdi = new tsdi_1.TSDI();
    });
    afterEach(function () {
        tsdi.close();
    });
    describe('when in mock mode', function () {
        beforeEach(function () {
            tsdi.enableComponentScanner();
        });
        it('should inject an auto-mocked instance', function () {
            var created = false;
            var Foo = /** @class */ (function () {
                function Foo() {
                }
                Foo.prototype.foo = function () {
                    created = true;
                };
                Foo = __decorate([
                    tsdi_1.component
                ], Foo);
                return Foo;
            }());
            var Bar = /** @class */ (function () {
                function Bar() {
                }
                Bar.prototype.init = function () {
                    this.foo.foo();
                };
                __decorate([
                    tsdi_1.inject,
                    __metadata("design:type", Foo)
                ], Bar.prototype, "foo", void 0);
                __decorate([
                    tsdi_1.initialize,
                    __metadata("design:type", Function),
                    __metadata("design:paramtypes", []),
                    __metadata("design:returntype", void 0)
                ], Bar.prototype, "init", null);
                Bar = __decorate([
                    tsdi_1.component
                ], Bar);
                return Bar;
            }());
            tsdi.enableAutomock();
            tsdi.get(Bar);
            chai_1.assert.isFalse(created);
        });
        it('should inject real instance if allowed', function () {
            var created = false;
            var Foo = /** @class */ (function () {
                function Foo() {
                }
                Foo.prototype.foo = function () {
                    created = true;
                };
                Foo = __decorate([
                    tsdi_1.component
                ], Foo);
                return Foo;
            }());
            var Bar = /** @class */ (function () {
                function Bar() {
                }
                Bar.prototype.init = function () {
                    this.foo.foo();
                };
                __decorate([
                    tsdi_1.inject,
                    __metadata("design:type", Foo)
                ], Bar.prototype, "foo", void 0);
                __decorate([
                    tsdi_1.initialize,
                    __metadata("design:type", Function),
                    __metadata("design:paramtypes", []),
                    __metadata("design:returntype", void 0)
                ], Bar.prototype, "init", null);
                Bar = __decorate([
                    tsdi_1.component
                ], Bar);
                return Bar;
            }());
            tsdi.enableAutomock(Foo);
            tsdi.get(Bar);
            chai_1.assert.isTrue(created);
        });
        it('should return a mock', function () {
            var created = false;
            var Foo = /** @class */ (function () {
                function Foo() {
                }
                Foo.prototype.foo = function () {
                    //
                };
                Foo = __decorate([
                    tsdi_1.component
                ], Foo);
                return Foo;
            }());
            var Bar = /** @class */ (function () {
                function Bar() {
                }
                Bar.prototype.init = function () {
                    this.foo.foo();
                };
                __decorate([
                    tsdi_1.inject,
                    __metadata("design:type", Foo)
                ], Bar.prototype, "foo", void 0);
                __decorate([
                    tsdi_1.initialize,
                    __metadata("design:type", Function),
                    __metadata("design:paramtypes", []),
                    __metadata("design:returntype", void 0)
                ], Bar.prototype, "init", null);
                Bar = __decorate([
                    tsdi_1.component
                ], Bar);
                return Bar;
            }());
            tsdi.enableAutomock();
            tsdi.mock(Foo).foo = function () { return created = true; };
            tsdi.get(Bar);
            chai_1.assert.isTrue(created);
        });
    });
});
