// @vsintellicode/vscode-intellicode-java/build/command.js
"use strict";
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
const vscode = __importStar(require("vscode"));
const JAVA_MODE = [
    { language: "java", scheme: "file" },
];
const JAVA_COMPLETION = "java.intellicode.completion";
const JAVA_COMPLETION_RESOLVE = "java.intellicode.completion.resolve";
const ENABLE_JAVA_INTELLICODE = "java.intellicode.enable";
const EXECUTE_WORKSPACE_COMMAND = "java.execute.workspaceCommand";
async function registerCompletionItemProvider(modelPath) {
    const result = await vscode.commands.executeCommand(EXECUTE_WORKSPACE_COMMAND, ENABLE_JAVA_INTELLICODE, true, modelPath);
    if (result) {
        return vscode.languages.registerCompletionItemProvider(JAVA_MODE, {
            provideCompletionItems(document, position, _token, completionContext) {
                return vscode.commands.executeCommand(EXECUTE_WORKSPACE_COMMAND, JAVA_COMPLETION, {
                    textDocument: { uri: document.uri.toString() },
                    position,
                    context: completionContext,
                });
            },
            resolveCompletionItem(item, _token) {
                return vscode.commands.executeCommand(EXECUTE_WORKSPACE_COMMAND, JAVA_COMPLETION_RESOLVE, item).then(convertCompletionItem);
            },
        }, ".", "@", "#", "*");
    }
    return false;
}
exports.registerCompletionItemProvider = registerCompletionItemProvider;
function convertCompletionItem(item) {
    if (item.textEdit) {
        try {
            item.insertText = new vscode.SnippetString(item.textEdit.newText);
            item.range = new vscode.Range(new vscode.Position(item.textEdit.range.start.line, item.textEdit.range.start.character), new vscode.Position(item.textEdit.range.end.line, item.textEdit.range.end.character));
            delete item.textEdit;
        }
        catch (error) {
            console.log(error);
        }
    }
    else if (item.insertTextFormat === 2) {
        item.insertText = new vscode.SnippetString(item.insertText);
    }
    return item;
}
