// @vsintellicode/vscode-intellicode-java/build/vscode-intellicode-java.js
"use strict";
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
const path = __importStar(require("path"));
const vscode = __importStar(require("vscode"));
const javaCommand = __importStar(require("./command"));
const minModelSchemaVersion = 1;
const maxModelSchemaVersion = 1;
let useLocalModel = false;
class JavaSupport {
    getRequestedConfig() {
        if (arguments.length > 0) {
            useLocalModel = arguments[1];
        }
        return [{
                scopeName: "vsintellicode",
                settingName: "java.completionsEnabled",
                desiredValue: true,
                required: true,
                scopesToTry: [
                    vscode.ConfigurationTarget.Global,
                    vscode.ConfigurationTarget.Workspace,
                    vscode.ConfigurationTarget.WorkspaceFolder,
                ],
                reloadWindowAfterApplying: false,
                notificationMessage: "IntelliCode Java support requires you to enable it.",
                actionLabel: "Enable it and Reload Window",
            }];
    }
    async activate(api, logger) {
        if (useLocalModel) {
            await javaCommand.registerCompletionItemProvider(path.join(__dirname, "../test/resources/MODEL"));
        }
        else {
            const model = await api.ModelAcquisitionService
                .getModelProvider("JavaIntellicode", "Completions", minModelSchemaVersion, maxModelSchemaVersion, [])
                .getModelAsync(undefined);
            if (model === undefined) {
                logger("No model available for Java, cannot continue.");
                return Promise.resolve();
            }
            const modelJson = JSON.stringify(model);
            logger(`vs-intellicode-java was passed a model: ${modelJson}.`);
            await javaCommand.registerCompletionItemProvider(model.modelPath);
            logger("Enabled IntelliCode for Java.");
            return Promise.resolve();
        }
    }
}
exports.JavaSupport = JavaSupport;
